infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesClientTermination.java

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright (c) 2012-2018 Red Hat, Inc.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.workspace.infrastructure.kubernetes;
+
+import javax.inject.Inject;
+import org.eclipse.che.api.system.server.ServiceTermination;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Termination for Kubernetes HTTP client.
+ *
+ * @author Max Shaposhnik (mshaposh@redhat.com)
+ */
+public class KubernetesClientTermination implements ServiceTermination {
```

Comment:
Is there any guarantee that `WorkspaceServiceTermination` will be invoked before `KubernetesClientTermination`?

Reply:
No :) But at least we can change binding order if something will go wrong :)

Reply:
It fits our needs better, but i see very weak logic how K8S client stop extends workspace shutdowning...